File: examples/demo-HomeMenuControl_es/src/com/example/hometest/MainActivity.java
package com.example.hometest;

import android.app.Activity;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.view.KeyEvent;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

public class MainActivity extends Activity {

	Button btnMenuY;
	Button btnMenuN;
	Button btnHomeY;
	Button btnHomeN;
	Button btndownY;
	Button btndownN;
	Button btntouchY;
	Button btntouchN;
	Button btnscreenY;
	Button btnscreenN;
	Button btnscapY;
	Button btnscapN;
	
	TextView tv;

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_main);

		HomeKeyEventBroadCastReceiver receiver = new HomeKeyEventBroadCastReceiver();
		registerReceiver(receiver, new IntentFilter(
				"com.rscja.android.KEY_DOWN"));

		InitView();
	}

	void InitView() {
		tv = (TextView) findViewById(R.id.textView4);
		btnMenuY = (Button) findViewById(R.id.btnmenuyes);
		btnMenuN = (Button) findViewById(R.id.btnmenuno);
		btnHomeY = (Button) findViewById(R.id.btnhomeyes);
		btnHomeN = (Button) findViewById(R.id.btnhomeno);
		btndownY = (Button) findViewById(R.id.btndownyes);
		btndownN = (Button) findViewById(R.id.btndownno);
		btntouchY = (Button) findViewById(R.id.btntouchY);
		btntouchN = (Button) findViewById(R.id.btntouchN);
		btnscreenY=(Button)findViewById(R.id.btnscreenY);
		btnscreenN=(Button)findViewById(R.id.btnscreenN);
		btnscapY=(Button)findViewById(R.id.btnscapY);
		btnscapN=(Button)findViewById(R.id.btnscapN);
		
		btnMenuY.setOnClickListener(new View.OnClickListener() {

			@Override
			public void onClick(View v) {
				// TODO Auto-generated method stub
				Intent intent = new Intent("android.intent.action.MENU_ENABLE");
				sendBroadcast(intent);
			}
		});

		btnMenuN.setOnClickListener(new View.OnClickListener() {

			@Override
			public void onClick(View v) {
				// TODO Auto-generated method stub
				Intent intent = new Intent(
						"android.intent.action.MENU_DISABLED");
				sendBroadcast(intent);
			}
		});

		btnHomeY.setOnClickListener(new View.OnClickListener() {

			@Override
			public void onClick(View v) {
				// TODO Auto-generated method stub
				Intent intent = new Intent("android.intent.action.HOME_ENABLE");
				sendBroadcast(intent);
			}
		});

		btnHomeN.setOnClickListener(new View.OnClickListener() {

			@Override
			public void onClick(View v) {
				// TODO Auto-generated method stub
				Intent intent = new Intent(
						"android.intent.action.HOME_DISABLED");
				sendBroadcast(intent);
			}
		});

		btndownY.setOnClickListener(new View.OnClickListener() {

			@Override
			public void onClick(View v) {
				// TODO Auto-generated method stub
				Intent intent = new Intent("android.intent.action.PANEL_ENABLE");
				sendBroadcast(intent);
			}
		});

		btndownN.setOnClickListener(new View.OnClickListener() {

			@Override
			public void onClick(View v) {
				// TODO Auto-generated method stub
				Intent intent = new Intent(
						"android.intent.action.PANEL_DISABLED");
				sendBroadcast(intent);
			}
		});
		
		btntouchY.setOnClickListener(new View.OnClickListener() {
			
			@Override
			public void onClick(View v) {
				// TODO Auto-generated method stub
				Intent intent = new Intent(
						"android.intent.action.TOUCH_ENABLE");
				sendBroadcast(intent);
			}
		});
		
		btntouchN.setOnClickListener(new View.OnClickListener() {
			
			@Override
			public void onClick(View v) {
				// TODO Auto-generated method stub
				Intent intent = new Intent(
						"android.intent.action.TOUCH_DISABLED");
				sendBroadcast(intent);
			}
		});

		btnscreenY.setOnClickListener(new View.OnClickListener() {
			
			@Override
			public void onClick(View v) {
				// TODO Auto-generated method stub
				Intent intent = new Intent(
						"android.intent.action.UnLock");
				sendBroadcast(intent);
			}
		});
		
		btnscreenN.setOnClickListener(new View.OnClickListener() {
			
			@Override
			public void onClick(View v) {
				// TODO Auto-generated method stub
				Intent intent = new Intent(
						"android.intent.action.Lock");
				sendBroadcast(intent);
			}
		});
		
		btnscapY.setOnClickListener(new View.OnClickListener() {
			
			@Override
			public void onClick(View v) {
				// TODO Auto-generated method stub
				Intent intent = new Intent(
						"android.intent.action.SCAP_ENABLE");
				sendBroadcast(intent);
			}
		});
		
		btnscapN.setOnClickListener(new View.OnClickListener() {
			
			@Override
			public void onClick(View v) {
				// TODO Auto-generated method stub
				Intent intent = new Intent(
						"android.intent.action.SCAP_DISABLED");
				sendBroadcast(intent);
			}
		});
	}

	@Override
	public boolean onKeyDown(int keyCode, KeyEvent event) {

		tv.setText(String.valueOf(keyCode));
		if(keyCode==139)
		{
			Intent intent = new Intent(
					"android.intent.action.TOUCH_ENABLE");
			sendBroadcast(intent);
			
		}
		return super.onKeyDown(keyCode, event);

	}

}

/*
 * 禁用功能后，此类监控真实键值，keydown事件不反映
 * 
 * After disable this function, the real key value of this class,keydown event
 * will not response.
 */

class HomeKeyEventBroadCastReceiver extends BroadcastReceiver {

	static final String SYSTEM_REASON = "reason";
	static final String SYSTEM_HOME_KEY = "homekey";// home key
	static final String SYSTEM_RECENT_APPS = "recentapps";// long home key

	@Override
	public void onReceive(Context context, Intent intent) {
		String action = intent.getAction();
		if (action.equals("com.rscja.android.KEY_DOWN")) {
			int reason = intent.getIntExtra("Keycode", 0);
			// getStringExtra
			boolean long1 = intent.getBooleanExtra("Pressed", false);

			// home key处理点
			Toast.makeText(
					context.getApplicationContext(),
					"home key=" + reason + ",long1="
							+ ((long1 == false) ? "Short Press" : "Long Press"),
					Toast.LENGTH_SHORT).show();
		}
	}

}

File: examples/demo-uhf_xm/UHF/MainActivity.cs
﻿using System;

using Android.App;
using Android.Content;
using Android.Runtime;
using Android.Views;
using Android.Widget;
using Android.OS;
using Com.Rscja.Deviceapi; 
using System.Collections; 
using System.Collections.Generic;
using Android.Util;

namespace UHF
{
	[Activity (Label = "UHF", MainLauncher = true)]
	public class MainActivity : Activity
	{
		public RFIDWithUHF uhfAPI;
		private ActionBar actionBar=null;
		private Scan_Fragment scan_Fragment=null;
		private Set_Fragment set_Fragment=null;
		private Write_Fragment write_Fragment =null;
		private Read_Fragment read_Fragment =null;
        HomeKeyEventBroadCastReceiver receiver;
        protected override void OnCreate (Bundle bundle)
		{
			base.OnCreate (bundle);
			// Set our view from the "main" layout resource
			SetContentView (Resource.Layout.Main);
			 
			 if (uhfAPI == null) {
				try
				{
					 uhfAPI=RFIDWithUHF.Instance;
				 	 uhfAPI.StopInventory ();
				}catch
				{
				
				}
			 }

            Log.Debug("DeviceAPI_",VersionInfo.Version+"--sssssss");

            initTab ();


            //receiver = new HomeKeyEventBroadCastReceiver(scan_Fragment);
            //RegisterReceiver(receiver, new IntentFilter("com.rscja.android.KEY_DOWN"));
        }

		protected override void OnResume()
		{
			base.OnResume ();
            Log.Info("re2",uhfAPI.ToString());
            if (uhfAPI != null) {
               
				new InitTask (this).Execute ();
			}
		}
		protected override void  OnPause()
		{
			base.OnPause ();
		}

        protected override void OnDestroy()
        {

            base.OnDestroy();
            uhfAPI.Free();
        }

        public override bool OnKeyDown(Keycode keyCode,KeyEvent e)
		{
           
			if (e.KeyCode.GetHashCode() == 139 || e.KeyCode.GetHashCode() == 280 || e.KeyCode.GetHashCode() == 293) {
				if (e.RepeatCount == 0) {
					if (actionBar.SelectedTab.Tag != null) {
						//if (actionBar.SelectedTab.Tag.ToString() == "Scan") {
							scan_Fragment.scan ();
                           
							return true;
						//}
					}
				}
			}
			return base.OnKeyDown (keyCode, e);


		}
		void initTab()
		{
			scan_Fragment = new Scan_Fragment ();
			set_Fragment = new Set_Fragment ();
			write_Fragment = new Write_Fragment ();
			read_Fragment = new Read_Fragment ();

			actionBar = ActionBar;
			actionBar.NavigationMode = ActionBarNavigationMode.Tabs;
			actionBar.SetDisplayHomeAsUpEnabled (true);

			ActionBar.Tab tab1 = actionBar.NewTab ();
			tab1.SetTag("Scan");
			tab1.SetText(Resource.String.tab_Scan);
			tab1.TabSelected += (object sender, ActionBar.TabEventArgs e) => {
				e.FragmentTransaction.Replace(Android.Resource.Id.Content,scan_Fragment);
			};   

			ActionBar.Tab tab2 = actionBar.NewTab ();
			tab1.SetTag ("Write");
			tab2.SetText (Resource.String.tab_WriteData);
			tab2.TabSelected += (object sender, ActionBar.TabEventArgs e) => {
				e.FragmentTransaction.Replace(Android.Resource.Id.Content,write_Fragment);
			};
			ActionBar.Tab tab3 = actionBar.NewTab ();
			tab1.SetTag ("Read");
			tab3.SetText (Resource.String.tab_ReadData);
			tab3.TabSelected += (object sender, ActionBar.TabEventArgs e) => {
				e.FragmentTransaction.Replace(Android.Resource.Id.Content,read_Fragment);
			};

			ActionBar.Tab tab4 = actionBar.NewTab ();
			tab1.SetTag ("Config");
			tab4.SetText (Resource.String.tab_set);
			tab4.TabSelected += (object sender, ActionBar.TabEventArgs e) => {
				e.FragmentTransaction.Replace(Android.Resource.Id.Content,set_Fragment);
			};
			actionBar.AddTab (tab1);
			actionBar.AddTab (tab2);
			actionBar.AddTab (tab3);
			actionBar.AddTab (tab4);
		}

		private class InitTask: AsyncTask<Java.Lang.Void, Java.Lang.Void, string[]>{

			MainActivity mainActivity;
			ProgressDialog  proDialg=null;

			public InitTask(MainActivity _mainActivity)
			{
				mainActivity=_mainActivity;
			}
			protected override string[] RunInBackground (params Java.Lang.Void[] @params)
			{
				//throw new NotImplementedException ();
				return null;
			}

			//后台要执行的任务
			protected override Java.Lang.Object DoInBackground(params Java.Lang.Object[] @params)
			{
				string result ="No";
				if  (mainActivity.uhfAPI!=null)  
				{
                    for (int i = 1; i <= 3; i++)
                    {
                        if (i != 3)
                        {
                            if (!mainActivity.uhfAPI.Init())
                                mainActivity.uhfAPI.Free();
                            else
                             return "OK";
                            
                        }
                        else
                        {
                            if (mainActivity.uhfAPI.Init())
                            {
                                result = "OK";
                            }
                        }
                       
                    
                    }
					
				}
				return result;
			}
			protected override void OnPostExecute(Java.Lang.Object result)
			{
				proDialg.Cancel();
                Log.Debug("结果", result.ToString());
				if (result.ToString() != "OK")
					Toast.MakeText (mainActivity,"Init failure!",ToastLength.Short);
			}

			//开始执行任务
			protected override void OnPreExecute()
			{
				proDialg=new ProgressDialog(mainActivity);
				proDialg.SetMessage("init.....");
				proDialg.Show();
			}
		}


	}


    class HomeKeyEventBroadCastReceiver : BroadcastReceiver
    {

        String SYSTEM_REASON = "reason";
        String SYSTEM_HOME_KEY = "homekey";//home key
        String SYSTEM_RECENT_APPS = "recentapps";//long home key
        Scan_Fragment scan;


        public HomeKeyEventBroadCastReceiver(Scan_Fragment scan_Fragment)
        {
            scan = scan_Fragment;
        }
        public override void OnReceive(Context context, Intent intent)
        {

            String action = intent.Action;// getAction();
            if (action.Equals("com.rscja.android.KEY_DOWN"))
            {
                int reason = intent.GetIntExtra("keycode", 0);
                if (reason == 0)
                    reason = intent.GetIntExtra("Keycode", 0);

                //getStringExtra
                bool long1 = intent.GetBooleanExtra("Pressed", false);
                // home key处理点
                // Toast.makeText(getApplicationContext(), "home key=" + reason + ",long1=" + long1, Toast.LENGTH_SHORT).show();
                if (reason == 139 || reason == 280)
                {

                    scan.scan();


                }
            }
        }
    }
}


